contracts/near/eth-client/src/lib.rs

```diff
@@ -416,6 +410,97 @@ impl EthClient {
         }
     }
 
+    #[cfg(feature = "bsc")]
+    fn bsc_record_header(&mut self, header: BlockHeader) {
+        env::log("Record header".as_bytes());
+        let best_info = self.infos.get(&self.best_header_hash).unwrap();
+        let header_hash = header.hash.unwrap();
+        let header_number = header.number;
+        if header_number + self.finalized_gc_threshold < best_info.number {
+            panic!("Header is too old to have a chance to appear on the canonical chain.");
+        }
+
+        let parent_info = self
+            .infos
+            .get(&header.parent_hash)
+            .expect("Header has unknown parent. Parent should be submitted first.");
+
+        // Record this header in `all_hashes`.
+        let mut all_hashes = self
+            .all_header_hashes
+            .get(&header_number)
+            .unwrap_or_default();
+        assert!(
+            all_hashes.iter().find(|x| **x == header_hash).is_none(),
+            "Header is already known. Number: {}",
+            header_number
+        );
+        all_hashes.push(header_hash);
+        self.all_header_hashes.insert(&header_number, &all_hashes);
+
+        env::log("Inserting header".as_bytes());
+        // Record full information about this header.
+        self.headers.insert(&header_hash, &header);
+        let info = HeaderInfo {
+            total_difficulty: parent_info.total_difficulty + header.difficulty,
+            parent_hash: header.parent_hash.clone(),
+            number: header_number,
+        };
+        self.infos.insert(&header_hash, &info);
+        env::log("Inserted".as_bytes());
+
+        let mut final_epoch_block = self.bsc_get_final_epoch_block();
+        let (_validators, validators_len) = EthClient::bsc_get_validator_set_from_block(&final_epoch_block);
+
+        let finality_threshold = validators_len / 2;
+        let num_of_blocks_to_final = header_number - best_info.number;
+
+        // Check if canonical chain needs to be updated.
+        if num_of_blocks_to_final >= finality_threshold {
+            env::log("Canonical chain needs to be updated.".as_bytes());
+            let mut current_block_hash = info.parent_hash;
+            let mut current_block_number = header_number - 1;
+            let last_final_block_hash = self.best_header_hash;
+            // Update the final epoch to cover the edge case when the update run the first time after init
+            if EthClient::bsc_is_epoch(best_info.number) {
+                self.final_epoch_header_hash = last_final_block_hash;
+                final_epoch_block = self.bsc_get_final_epoch_block();
+            }
```

Comment:
If it's related only to the first time init, let's add another flag to make sure that:
* it won't be called in all other cases
* it won't waste the gas every time

Alternatively, we can require adding epoch block header and `epoch_block_header.number - 1` and so we will get rid of these special conditions. We can do this as we don't expect to run the client from block 0 and remove this requirement in one of previous PRs.

Reply:
fixed 56af8eaeed9a8b15b050cdbd32cec7b3d36b688c